pkg/controller/namespace/namespace_controller_utils.go

```diff
@@ -264,125 +327,76 @@ func syncNamespace(kubeClient clientset.Interface, versions *unversioned.APIVers
 	return nil
 }
 
-func deleteLimitRanges(kubeClient clientset.Interface, ns string) error {
-	return kubeClient.Core().LimitRanges(ns).DeleteCollection(nil, api.ListOptions{})
-}
-
-func deleteResourceQuotas(kubeClient clientset.Interface, ns string) error {
-	return kubeClient.Core().ResourceQuotas(ns).DeleteCollection(nil, api.ListOptions{})
-}
-
-func deleteServiceAccounts(kubeClient clientset.Interface, ns string) error {
-	return kubeClient.Core().ServiceAccounts(ns).DeleteCollection(nil, api.ListOptions{})
-}
-
-func deleteServices(kubeClient clientset.Interface, ns string) error {
-	items, err := kubeClient.Core().Services(ns).List(api.ListOptions{})
+// estimateGrracefulTermination will estimate the graceful termination required for the specific entity in the namespace
+func estimateGracefulTermination(kubeClient clientset.Interface, groupVersionResource unversioned.GroupVersionResource, ns string, namespaceDeletedAt unversioned.Time) (int64, bool, error) {
+	groupResource := groupVersionResource.GroupResource()
+	glog.V(4).Infof("namespace controller - estimateGracefulTermination - group %s, resource: %s", groupResource.Group, groupResource.Resource)
+	estimate := int64(0)
+	supportsGrace := false
+	var err error
+	switch groupResource {
+	case unversioned.GroupResource{Group: "", Resource: "pods"}:
+		estimate, supportsGrace, err = estimateGracefulTerminationForPods(kubeClient, ns)
+	}
 	if err != nil {
-		return err
+		return estimate, supportsGrace, err
 	}
-	for i := range items.Items {
-		err := kubeClient.Core().Services(ns).Delete(items.Items[i].Name, nil)
-		if err != nil && !errors.IsNotFound(err) {
-			return err
-		}
+	// determine if the estimate is greater than the deletion timestamp
+	duration := time.Since(namespaceDeletedAt.Time)
+	allowedEstimate := time.Duration(estimate) * time.Second
+	if duration >= allowedEstimate {
+		estimate = int64(0)
 	}
-	return nil
+	return estimate, supportsGrace, nil
 }
 
-func deleteReplicationControllers(kubeClient clientset.Interface, ns string) error {
-	return kubeClient.Core().ReplicationControllers(ns).DeleteCollection(nil, api.ListOptions{})
-}
-
-func deletePods(kubeClient clientset.Interface, ns string, before unversioned.Time) (int64, error) {
+// estimateGracefulTerminationForPods determines the graceful termination period for pods in the namespace
+func estimateGracefulTerminationForPods(kubeClient clientset.Interface, ns string) (int64, bool, error) {
+	glog.V(4).Infof("namespace controller - estimateGracefulTerminationForPods - namespace %s", ns)
 	items, err := kubeClient.Core().Pods(ns).List(api.ListOptions{})
 	if err != nil {
-		return 0, err
-	}
-	expired := unversioned.Now().After(before.Time)
-	var deleteOptions *api.DeleteOptions
-	if expired {
-		deleteOptions = api.NewDeleteOptions(0)
+		return 0, true, err
 	}
 	estimate := int64(0)
 	for i := range items.Items {
+		// filter out terminal pods
```

Comment:
this is new in this PR and was intended to fix:
https://github.com/kubernetes/kubernetes/issues/21701

If a pod is in a terminal state, there is no needed to unnecessarily wait for graceful deletion of that pod because its already terminated.